res/controllers/Numark-Mixtrack-Pro-FX-scripts.js

```diff
@@ -0,0 +1,782 @@
+            }
+            return false;
+        }
+    });
+
+    this.loop = new components.Button({
+        outKey: "loop_enabled",
+        midi: [0x94 + channel, 0x40],
+        shift: function() {
+            this.inKey = "loop_in_goto";
+            this.input = components.Button.prototype.input;
+        },
+        unshift: function() {
+            this.input = function(channel, control, value, status, group) {
+                if (value !== 0x7F) {
+                    return;
+                }
+                if (engine.getValue(group, "loop_enabled") === 0) {
+                    script.triggerControl(group, "beatloop_activate");
+                } else {
+                    script.triggerControl(group, "beatlooproll_activate");
+                }
+            };
+        },
+        shiftControl: true,
+        sendShifted: true,
+        shiftOffset: 0x01
+    });
+
+    this.loopHalf = new components.Button({
+        midi: [0x94 + channel, 0x34],
+        shiftControl: true,
+        sendShifted: true,
+        shiftOffset: 0x02,
+        shift: function() {
+            this.inKey = "loop_in";
+            this.outKey = "loop_in";
+            this.disconnect();
+            this.connect();
+        },
+        unshift: function() {
+            this.inKey = "loop_halve";
+            this.outKey = "loop_halve";
+            this.disconnect();
+            this.connect();
+        }
+    });
+
+    this.loopDouble = new components.Button({
+        midi: [0x94 + channel, 0x35],
+        shiftControl: true,
+        sendShifted: true,
+        shiftOffset: 0x02,
+        shift: function() {
+            this.inKey = "loop_out";
+            this.outKey = "loop_out";
+            this.disconnect();
+            this.connect();
+        },
+        unshift: function() {
+            this.inKey = "loop_double";
+            this.outKey = "loop_double";
+            this.disconnect();
+            this.connect();
+        }
+    });
+
+    this.scratchToggle = new components.Button({
+        midi: [0x90 + channel, 0x07],
+        shift: function() {
+            this.inKey = "reverseroll";
+            this.outKey = "reverseroll";
+            this.connect();
+            this.input = components.Button.prototype.input;
+        },
+        unshift: function() {
+            this.disconnect();
+            this.input = function(channel, control, value) {
+                if (value !== 0x7F) {
+                    return;
+                }
+                deck.scratchModeEnabled = !deck.scratchModeEnabled;
+                midi.sendShortMsg(this.midi[0], this.midi[1], deck.scratchModeEnabled ? 0x7F : 0x01);
+            };
+
+            midi.sendShortMsg(this.midi[0], this.midi[1], deck.scratchModeEnabled ? 0x7F : 0x01);
+
+            var bleepEnabled = engine.getParameter(this.group, "reverseroll") === 1;
+            midi.sendShortMsg(this.midi[0], this.midi[1] + this.shiftOffset, bleepEnabled ? 0x7F : 0x01);
```

Comment:
is this even required if you just don't disconnect the button at the beginning of the function? So always have the outwards LED binding and only overwrite the input function in the (un-)shift methods?

Reply:
It almost works, but breaks in the following scenario:
1. "scratch mode" is enabled (light on)
2. Shift down
3. Scratch button down
4. Shift up
5. Scratch button up

Then the button ends up with light disabled (0x01), because reverseroll has just been disabled (by releasing the button). It should have light enabled because "scratch mode" is still enabled.

Reply:
I see. I guess just keep it as is then. It should be fine since its only a very local piece of code. Maybe add some comments explaining what everything does.